### PR TITLE
Add withSafeInterval HOC

### PR DESCRIPTION
**Pull request by tfrommen**
## Description

This pull request adds and applies a new higher-order component (HOC), `withSafeInterval`, which works just along the lines of the existing `withSafeTimeout` HOC.

## How has this been tested?

I ran `npm run test-unit`.

## Types of changes

* Add new HOC to wrap `window.setInterval`.
* Use new HOC where `window.setInterval` was used before.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/compose/src/with-safe-interval/index.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+			}
+
+			componentWillUnmount() {
+				this.intervals.forEach( clearInterval );
+			}
+
+			setInterval( fn, delay ) {
+				const id = setInterval( () => {
+					fn();
+					this.clearInterval( id );
```
Comment (aduth) [inline packages/compose/src/with-safe-interval/index.js]: It's not clear to me: Why would the default behavior clear the interval after the initial tick?

Comment (tfrommen) [reply packages/compose/src/with-safe-interval/index.js]: Good catch! It shouldn't. This was a left-over from the `withSafeTimeout` file, which I copied and then adjusted. 👍

**Comment by aduth** [comment]
Are you able to take a stab at writing some unit tests for this?

File: packages/components/src/popover/index.js [hunk 2]
```diff
@@ -95,6 +96,8 @@ class Popover extends Component {
 	}
 
 	toggleAutoRefresh( isActive ) {
+		const { clearInterval, setInterval } = this.props;
```
Comment (aduth) [inline packages/components/src/popover/index.js]: It begs the question: If the component is taking advantage of `withSafeInterval`, does it really need to call `clearInterval` itself? As I can tell, the `clearInterval` would only be called as a result of `componentWillUnmount`, the same which would be handled by `withSafeIntervals`.

Comment (tfrommen) [reply packages/components/src/popover/index.js]: Hm, I'm not following. The existing code calls `clearInterval` [here](https://github.com/WordPress/gutenberg/pull/14818/files#diff-ccdf667da5123277e79748286c32fda7R118). So I'd say, yes, there is still the need to stop the callback from executing. Or did I miss something?

Comment (aduth) [reply packages/components/src/popover/index.js]: The purpose of a `withSafeInterval` is that it handles this on behalf of the component:

https://github.com/WordPress/gutenberg/pull/14818/files#diff-fe5750aa4c4751d70bcf38cb66604118R35

The existing code didn't use `withSafeInterval`, so it needed to call it itself.

**Comment by aduth** [comment]
With the advent of [React hooks](https://reactjs.org/docs/hooks-intro.html), i wonder if this is something we still need, and if it were to exist, whether it should be written as a hook. There's an argument to be made of offering this merely for consistency's sake.

Example React hooks usage:

```js
function MyComponent() {
	useEffect( () => {
		const intervalId = setInterval( () => {}, 1000 );
		return () => clearInterval( intervalId );
	}, [] );

	// ...
}
```

As a custom hook:

```js
const useInterval = ( callback, interval ) => useEffect( () => {
	const intervalId = setInterval( callback, interval );
	return () => clearInterval( intervalId );
}, [] );

function MyComponent() {
	useInterval( () => {}, 1000 );

	// ...
}
```

**Pull request closed without merging**